v3/mathjax3-ts/core/MathDocument.ts

```diff
@@ -0,0 +1,401 @@
+ *
+ *  Copyright (c) 2017 The MathJax Consortium
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+/**
+ * @fileoverview  Implements the interface and abstract class for MathDocument objects
+ *
+ * @author dpvc@mathjax.org (Davide Cervone)
+ */
+
+import {userOptions, defaultOptions, OptionList} from '../util/Options.js';
+import {InputJax, AbstractInputJax} from './InputJax.js';
+import {OutputJax, AbstractOutputJax} from './OutputJax.js';
+import {MathList, MathListClass, AbstractMathList} from './MathList.js';
+import {MathItem, AbstractMathItem} from './MathItem.js';
+
+/*****************************************************************/
+/*
+ *  The MathDocument interface
+ *
+ *  The MathDocument is created by MathJax.Document() and holds the
+ *  document, the math found in it, and so on.  The methods of the
+ *  MathDocument all return the MathDocument itself, so you can
+ *  chain the method calls.  E.g.,
+ *
+ *    const html = MathJax.Document('<html>...</html>');
+ *    html.findMath()
+ *        .compile()
+ *        .getMetrics()
+ *        .typeset()
+ *        .addEventHandlers()
+ *        .updateDocument();
+ *
+ *  The MathDocument is the main interface for page authors to
+ *  interact with MathJax.
+ */
+
+export interface MathDocument {
+    /*
+     * The document being processed (e.g., DOM document, or Markdown string)
+     */
+    document: any;
+
+    /*
+     * The kind of MathDocument (e.g., "HTML")
+     */
+    kind: string;
+
+    /*
+     * The options for the document
+     */
+    options: OptionList;
+
+    /*
+     * The list of MathItems found in this page
+     */
+    math: MathList;
+
+    /*
+     * This object tracks what operations have been performed, so that (when
+     *  asynchronous operations are used), the ones that have already been
+     *  completed won't be performed again.
+     */
+    processed: {[name: string]: boolean};
+
+    /*
+     * An array of input jax to run on the document
+     */
+    inputJax: InputJax[];
+
+    /*
+     * The output jax to use for the document
+     */
+    outputJax: OutputJax;
+
+    /*
+     * Locates the math in the document and constructs the MathList
+     *  for the document.
+     *
+     * @param{OptionList} options  The options for locating the math
+     * @return{MathDocument}       The math document instance
+     */
+    findMath(options?: OptionList): MathDocument;
```

Comment:
It is more complicated than that.  For one thing, the handler has to create the array of strings from the document for the FindMath object to search, so it is the handler that would call FindMath.  See our email for more.